### DISCUSSION TITLE
MaxLFQ by diann R package

### DISCUSSION THREAD

**Discussion by JM-Bader** [vdemichev/DiaNN#352] [Q&A] [answered]
https://github.com/vdemichev/diann-rpackage

The DIA-NN R package can be used to apply the MaxLFQ algorithm after custom batch correction. 
Unfortunately, the example workflow on github does not include a batch correction step, so it is not completely clear how to fit that in. 

Q1: What needs to be inputted into the package for MaxLFQ normalisation ? Is it the entire report after batch correction? ("df <- diann_load("diann_report.tsv")")?
Q2: I presume it is advantageous to filter for 1% FDR before batch correction, correct?
Q3: What do you recommend for batch correction?
Q4: What is your stance on imputation? If yes, imputation on precursors level before MaxLFQ ? Or after MaxLFQ on protein group level?
Q5: There are two lines on non-modified peptides.

Peptides without modifications - taking the maximum of the respective precursor quantities

peptides <- diann_matrix(df, id.header="Stripped.Sequence", pg.q = 0.01)

Peptides without modifications - using the MaxLFQ algorithm

peptides.maxlfq <- diann_maxlfq(df[df$Q.Value <= 0.01 & df$PG.Q.Value <= 0.01,], group.header="Stri

What is the context? Shall one exclude modified peptides? Why ?
Is it possible to use all peptides (modified and non-modified) for batch correction and subsequent MaxLFQ?

Q6: As a suggestion: It would be nice to load a table (in a given format) into DIA-NN with annotation of batches and then let DIA-NN do batch correction on precursor level and subsequent MaxLFQ. Is that planned ?

Q7: are there data (publication) or in your experience - how does batch correction on precursor level compare to batch correction on protein level?

**Comment by vdemichev** [comment, marked as answer]
Q1. diann_maxlfq summarises precursor quantities into protein quantities. The examples are provided. The columns required can be specified, see function specification by typing '?diann_maxlfq' in R.
Q2. Basic filtering is precursor level 1% run-specific and global + protein 1% global. Everything else is optional.
Q3. Please see our covid papers for some outline of how we do batch correction, as well as large-scale studies by others. But this is not a universal way, more fancy batch correction strategies are definitely likely to be better.
Q4. We usually either don't impute or use minimal-value imputation (across the profile of each protein). For machine learning we sometimes used KNN.
Q5. It's quantifying peptides using stripped sequence as the key, i.e. summarising all peptidoforms into a single quantity.
Q6. Yes, at some point.
Q7. We don't really have comprehensive information. We used to correct on precursor-level, but protein-level can also be OK, and some times it might be beneficial to use both I guess.

**Reply by JM-Bader** [reply]
Thanks for the answers above. I have more questions now: 
Q8: Is there a separate discussion board for the MaxLFQ R script? (other than this one supposed to be for DIANN)

Before using the MaxLFQ script for real applications I tried to check it and noticed a few things. I loaded the .tsv table from a DIA-NN search with the full information, performed MaxLFQ on it according to the instructions on the github page, and then compared the protein level output to the output found in the .pg.matrix.tsv. 
Q09: The MaxLFQ script output table column order (i.e. order of raw files) is partially messed up and does not correspond to either the .pg.matrix.tsv. raw file order nor to the order of raw files in the full df.tsv (using unique(df$Run)). 
Q10: The MaxLFQ script output table does not include the text columns "Protein.Group", "Protein.IDs", "Protein.Names"  "Genes", "First.Protein.Description". Just matching the protein group information back from the full .tsv may could be a severe issue if the order is messed up too (e.g. in light of the altered order of raw files - Q09). 
> My suggestion is to include the protein group matching (and the other text columns found in .pg.matrix.tsv) in the MaxLFQ script
Q11: The MaxLFQ script output table has 6% more rows (ie. protein groups) than the  .pg.matrix.tsv. What is the cause of that ? I used the default settings for both the DIA-NN GUI search (creating the .pg.matrix.tsv) and for the MaxLFQ script according to the github (creating the script output protein table from the full GUI search .tsv table). If the difference is inherent, what do you recommend? Otherwise, how do I have to set a key parameter to reach consistency?

Q12: What is the default margin setting (explained in the MaxLFQ script to be a intensity cutoff) in the DIA-NN GUI? The default value in the MaxLFQ script appears to be -10. How are negative intensities possible?

**Reply by vdemichev** [reply]
Q8 - no.
Q9 - this is OK. No particular order is guaranteed, can sort the files yourself in whatever preferred way - this is one line of R code.
Q10 - matching again should be easy regardless of raw files, as what is being matched is Protein.Group (or whatever is being quantified) - things like Protein.Ids, names, genes, etc - they are all the same for all Protein.Group entries in the main report, regardless of the run these entries correspond to.
Q11 - different numbers of rows mean different filtering applied., see https://github.com/vdemichev/DiaNN#output for the filtering of matrices. On consistency: we ourselves always do eveyrthing from the main report, which gives us full control over filtering the data the way we want. The 'matrices' are a convenient 'extra', with primary purpose to simplify things for those who only use MS Excel. So if any doubts about how they are created, please just use the main report. The filtering used to create matrices is the best for many but not all experiments.
Q12 - MaxLFQ implementation in DIA-NN itself is different. In the script it referes to log2 intensity (thus negative) and it's not supposed to be adjusted by the user.

**Reply by JM-Bader** [reply]
thanks for the swift reply. I am sorry, it appears we had a misunderstanding regarding Q10:
Q10: Clearly, matching of extra protein information does not depend on the columns. However, it either requires a reference ID column (e.g. Protein.Group which needs to be present in both tables) or an identical order of rows (i.e. Protein.Groups). There apparently is no reference ID column (e.g Protein.Group) in the MaxLFQ script output - only intensities. 
So is one supposed to match Protein.Group information based on the row order ? From which table? Could it be that the row order is different in the MaxLFQ output table ?

**Reply by vdemichev** [reply]
diann_maxlfq produces a matrix wherein rows are named - these names can be used for matching. 
A note: you only need diann_maxlfq function if you do precursor-level batch correction, otherwise can use DIA-NN-calculated MaxLFQ quantities, and summarise them in a table with diann_matrix.

**Reply by tvpham** [reply]
Hi @JM-Bader, I am also exploring DIA-NN output using an R package called iq (I am the author of the package), https://github.com/tvpham/iq . It can perform protein quantitation at the fragment level. Protein quantification at precursor level is possible too. The default setting is to filter run-specific FDR of 1% for both precursor and protein group. However, I am not sure if this is what most people do. Perhaps you could give it a try.
